### PR TITLE
feat: #954 add custom max chunk size to client put and putCar options

### LINKED ISSUE #954: Can I change the default chuck size?
Default chunk size is about 10MB, So the progress won't change for a long time,  is there any way I can change the chunk size so I can let the progress change more frequently. thanks

### PR DESCRIPTION
**Pull request by akazwz**
### Changes

-   add custom chunk size to client put and putCar options: fixes #954

### CODE REVIEW

File: packages/client/src/lib.js [hunk 1]
```diff
@@ -109,12 +110,12 @@ class Web3Storage {
         })),
         blockstore,
         wrapWithDirectory,
-        maxChunkSize: 1048576,
+        maxChunkSize: maxChunkSize,
```
Comment (alanshaw) [inline packages/client/src/lib.js:113]:
```suggestion
        maxChunkSize,
```

File: packages/client/src/lib/interface.ts [hunk 2]
```diff
@@ -101,6 +101,10 @@ export type PutOptions = {
    * Maximum times to retry a failed upload. Default: 5
    */
   maxRetries?: number
+  /**
+   * Maximum chunk size to upload. Default: 1024 * 1024 * 10 (in bytes)
```
Comment (alanshaw) [inline packages/client/src/lib/interface.ts:105]:
```suggestion
   * Maximum chunk size to upload in bytes. Default: 10,485,760
```

File: packages/client/src/lib/interface.ts [hunk 3]
```diff
@@ -138,6 +142,10 @@ export type PutCarOptions = {
    * Maximum times to retry a failed upload. Default: 5
    */
    maxRetries?: number
+  /**
+   * Maximum chunk size to upload. AS the targetSize as passed to carbites TreewalkSplitter. Default: 1024 * 1024 * 10 (in bytes).
```
Comment (alanshaw) [inline packages/client/src/lib/interface.ts:146]:
```suggestion
   * Maximum chunk size to upload in bytes. Default: 10,485,760
```

**Review by alanshaw — APPROVED** [review]

**Comment by alanshaw** [comment]
@vasco-santos @olizilla are you ok with this addition?

**Comment by vasco-santos** [comment]
> @vasco-santos @olizilla are you ok with this addition?

We should probably link to recommended values, or have a ceiling. Is there any benchmarks with different chunk sizes you are aware of?

File: packages/client/src/lib.js [hunk 4]
```diff
@@ -109,12 +110,12 @@ class Web3Storage {
         })),
         blockstore,
         wrapWithDirectory,
-        maxChunkSize: 1048576,
+        maxChunkSize,
```
Comment (alanshaw) [inline packages/client/src/lib.js:113]: Hold up, this shouldn't have been changed - this is the max block size (I know confusingly named) which we set to 1MB.

```suggestion
        maxChunkSize: 1048576,
```

Although we should really pull this out into a constant and use it here, so we don't get confused again!:

```js
const MAX_BLOCK_SIZE = 1048576
```

**Review by alanshaw — CHANGES_REQUESTED** [review]
@akazwz there's one fix that I've highlighted. The addition needed here is to add some validation on the `maxChunkSize` option to make sure it is less than 100MiB and greater than or equal to 1MB (`MAX_BLOCK_SIZE`).

File: packages/client/src/lib.js (hunk 4 above)
Comment (akazwz) [reply packages/client/src/lib.js]: @alanshaw Thanks for the suggestions, I already did some validation on the `maxChunkSize `and added the MAX_BLOCK_SIZE constant, have a look when you have time

File: packages/client/src/lib.js [hunk 5]
```diff
@@ -97,9 +98,13 @@ class Web3Storage {
     onRootCidReady,
     onStoredChunk,
     maxRetries = MAX_PUT_RETRIES,
+    maxChunkSize = MAX_CHUNK_SIZE,
     wrapWithDirectory = true,
     name
   } = {}) {
+    if (maxChunkSize >= 104857600 || maxChunkSize < 1048576) {
```
Comment (vasco-santos) [inline packages/client/src/lib.js:105]: @akazwz now that we have constant, we should use them in this conditional and other below. Also, we should have a constant for the 100MB upper limit

**Review by vasco-santos — CHANGES_REQUESTED** [review]
Thanks for the PR, left a comment about using the constants everywhere instead of still hardcoding the numbers

**Review by alanshaw — APPROVED** [review]

**Review by vasco-santos — APPROVED** [review]

**Review by vasco-santos — CHANGES_REQUESTED** [review]
@akazwz can you please add a test to get the changes validated and code coverage CI happy?

**Comment by alanshaw** [comment]
Ah we need tests for the coverage to pass!

**Comment by akazwz** [comment]
@alanshaw @vasco-santos Already added tests for custom chunk size, but I am not familiar with the tests, can you please check if it works or give some suggestions.

File: packages/client/test/put.spec.js [hunk 6]
```diff
@@ -169,6 +207,17 @@ describe('putCar', () => {
     }
   })
 
+  it('errors for CAR with wrong max chunk size', async () => {
+    const client = new Web3Storage({ token, endpoint })
+    const carReader = await createCar('hello world')
+    try {
+      await client.putCar(carReader, {maxChunkSize: 10})
```
Comment (vasco-santos) [inline packages/client/test/put.spec.js:214]:
```suggestion
      await client.putCar(carReader, { maxChunkSize: 10 })
```

**Review by vasco-santos — CHANGES_REQUESTED** [review]
Thanks a lot @akazwz this should be the last change needed

**Comment by akazwz** [comment]
@vasco-santos Already done, thanks for the suggestions

**Review by vasco-santos — APPROVED** [review]

**Pull request merged**